[PyTorch] Extend flatbuffer to support extra files map

pavithranrao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #72951

Extend flatbuffer to support extra files map

Flatbuffer schema has extra files. The users can write extra files by providing a `map<string, string>` which will be part of the flatbuffer model asset and and can be loaded back similar to pickle.

Differential Revision: [D34286346](https://our.internmc.facebook.com/intern/diff/D34286346/)

pavithranrao committed:
[PyTorch] Extend flatbuffer to support extra files map

Extend flatbuffer to support extra files map

Flatbuffer schema has extra files. The users can write extra files by providing a `map<string, string>` which will be part of the flatbuffer model asset and and can be loaded back similar to pickle.

Differential Revision: [D34286346](https://our.internmc.facebook.com/intern/diff/D34286346/)

[ghstack-poisoned]

pavithranrao committed:
Update on "[PyTorch] Extend flatbuffer to support extra files map"


Extend flatbuffer to support extra files map

Flatbuffer schema has extra files. The users can write extra files by providing a `map<string, string>` which will be part of the flatbuffer model asset and and can be loaded back similar to pickle.

Differential Revision: [D34286346](https://our.internmc.facebook.com/intern/diff/D34286346/)

[ghstack-poisoned]

qihqi commented on torch/csrc/jit/mobile/flatbuffer_loader.cpp:
```diff
@@ -172,6 +173,12 @@ mobile::Module FlatbufferLoader::parseModule(
   }
   IValue& module_ivalue = getIValue(module->state_obj());
 
+  auto extra_files_offsets = module_->extra_files();
```
Let's pull this out as a separate function. i.e. parseModule should only parseModule; if you want extra files call a separater function that returns the file content.

qihqi commented on torch/csrc/jit/mobile/flatbuffer_loader.h:
```diff
@@ -15,6 +15,7 @@
 namespace torch {
 namespace jit {
 
+using ExtraFilesMap = std::unordered_map<std::string, std::string>;
```
if there an ExtraFilesMap definition somewhere else? If so reuse that one.

pavithranrao replied:
It is not available while importing in mobile
 So it is redefined in [mobile/import.h](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/mobile/import.h#L14)
  
Once we integrate flatbuffer we can reuse the definition from mobile/import

qihqi replied:
Sorry why cannot this file #include <torch/csrc/jit/mobile/import.h>? import.h is not part of mobile?

pavithranrao committed:
Update on "[PyTorch] Extend flatbuffer to support extra files map"


Extend flatbuffer to support extra files map

Flatbuffer schema has extra files. The users can write extra files by providing a `map<string, string>` which will be part of the flatbuffer model asset and and can be loaded back similar to pickle.

Differential Revision: [D34286346](https://our.internmc.facebook.com/intern/diff/D34286346/)

[ghstack-poisoned]

pavithranrao committed:
Update on "[PyTorch] Extend flatbuffer to support extra files map"


Extend flatbuffer to support extra files map

Flatbuffer schema has extra files. The users can write extra files by providing a `map<string, string>` which will be part of the flatbuffer model asset and and can be loaded back similar to pickle.

Differential Revision: [D34286346](https://our.internmc.facebook.com/intern/diff/D34286346/)

[ghstack-poisoned]

qihqi commented on torch/csrc/jit/mobile/flatbuffer_loader.cpp:
```diff
@@ -172,6 +183,8 @@ mobile::Module FlatbufferLoader::parseModule(
   }
   IValue& module_ivalue = getIValue(module->state_obj());
 
+  parseExtraFiles(module, extra_files);
```
Please expose this function as something users can call. and not call it here.

qihqi commented on torch/csrc/jit/mobile/flatbuffer_loader.cpp:
```diff
@@ -543,22 +556,36 @@ TypePtr FlatbufferLoader::getOrCreateTypeAnnotations(
     mobile::serialization::Module* flatbuffer_module,
-    c10::optional<at::Device>) {
-  mobile::Module m = FlatbufferLoader().parseModule(flatbuffer_module);
+    ExtraFilesMap& extra_files,
```
let's leave those function unchanged. Let user call a separate function for those who need extra files.

qihqi commented on test/cpp/jit/test_flatbuffer.cpp:
```diff
@@ -137,6 +137,37 @@ TEST(FlatbufferTest, MethodInvocation) { // NOLINT (use =delete in gtest)
+  loaded_extra_files["metadata.json"] = "";
+  auto* flatbuffer_module =
+      mobile::serialization::GetMutableModule(buff.data());
+  mobile::Module bc2 =
```
this test no longer need to create a mobile::Module. We can call parseExtraFiles directly.

qihqi commented on torch/csrc/jit/mobile/flatbuffer_loader.h:
```diff
@@ -39,13 +45,19 @@ TORCH_API mobile::Module initialize_mobile_module(
 // versions above.
+TORCH_API mobile::Module load_mobile_module_from_file(
+    const std::string& filename,
+    ExtraFilesMap& extra_files,
```
Remove this param. 

User who need extra file shall call parseExtraFiles separately.

iseeyuan reviewed (approved):
LGTM. Please address @qihqi 's comments before landing. 
To make the migration smooth, Please make sure that the at the top level APIs not changed, including the way to save/load extra files:

Python
Module._save_for_lite_interpreter(self, path), and _save_to_buffer_for_lite_interpreter  for saving and
_load_for_lite_interpreter(path_or_stream) for loading.
C++
jit::Module::_save_for_mobile(path) for saving and
_load_for_mobile(path_or_stream) for loading.

pavithranrao committed:
Update on "[PyTorch] Extend flatbuffer to support extra files map"


Extend flatbuffer to support extra files map

Flatbuffer schema has extra files. The users can write extra files by providing a `map<string, string>` which will be part of the flatbuffer model asset and and can be loaded back similar to pickle.

Differential Revision: [D34286346](https://our.internmc.facebook.com/intern/diff/D34286346/)

[ghstack-poisoned]
